Sources/AppleAPI/Client.swift

```diff
@@ -87,24 +90,209 @@ public class Client {
+                    .validateSecurityCodeResponse()
+                    
+            }
+            .then { (data, response) -> Promise<Void>  in
+                self.updateSession(serviceKey: serviceKey, sessionID: sessionID, scnt: scnt)
+            }
         }
-        .then { (data, response) -> Promise<Void> in
-            self.session.dataTask(.promise, with: URLRequest.olympusSession).asVoid()
+    }
+    
+    func updateSession(serviceKey: String, sessionID: String, scnt: String) -> Promise<Void> {
+        return Current.network.dataTask(with: URLRequest.trust(serviceKey: serviceKey, sessionID: sessionID, scnt: scnt))
+            .then { (data, response) -> Promise<Void> in
+                Current.network.dataTask(with: URLRequest.olympusSession).asVoid()
+            }
+    }
+    
+    func selectPhoneNumberInteractively(from trustedPhoneNumbers: [AuthOptionsResponse.TrustedPhoneNumber]) -> Promise<AuthOptionsResponse.TrustedPhoneNumber> {
+        return firstly { () throws -> Guarantee<AuthOptionsResponse.TrustedPhoneNumber> in
+            Current.logging.log("Trusted phone numbers:")
+            trustedPhoneNumbers.enumerated().forEach { (index, phoneNumber) in
+                Current.logging.log("\(index + 1): \(phoneNumber.numberWithDialCode)")
+            }
+
+            let possibleSelectionNumberString = Current.shell.readLine("Select a trusted phone number to receive a code via SMS: ")
+            guard
+                let selectionNumberString = possibleSelectionNumberString,
+                let selectionNumber = Int(selectionNumberString) ,
+                trustedPhoneNumbers.indices.contains(selectionNumber - 1)
+            else {
+                throw Error.invalidPhoneNumberIndex(min: 1, max: trustedPhoneNumbers.count, given: possibleSelectionNumberString)
+            }
+
+            return .value(trustedPhoneNumbers[selectionNumber - 1])
+        }
+        .recover { error throws -> Promise<AuthOptionsResponse.TrustedPhoneNumber> in
+            guard case Error.invalidPhoneNumberIndex = error else { throw error }
+            Current.logging.log("\(error.localizedDescription)\n")
+            return self.selectPhoneNumberInteractively(from: trustedPhoneNumbers)
+        }
+    }
+    
+    func promptForSMSSecurityCode(length: Int, for trustedPhoneNumber: AuthOptionsResponse.TrustedPhoneNumber) -> SecurityCode {
+        let code = Current.shell.readLine("Enter the \(length) digit code sent to \(trustedPhoneNumber.numberWithDialCode): ") ?? ""
+        return .sms(code: code, phoneNumberId: trustedPhoneNumber.id)
+    }
+    
+    func handleWithPhoneNumberSelection(authOptions: AuthOptionsResponse, serviceKey: String, sessionID: String, scnt: String) -> Promise<Void> {
+        return firstly {
+            selectPhoneNumberInteractively(from: authOptions.trustedPhoneNumbers!)
```

Comment:
is it guaranteed that there will always be trusted phone numbers?

Reply:
Good question, this'll now throw an appropriate error if trustedPhoneNumbers is nil or empty. I don't think this should be possible because [2FA requires at least one trusted phone number](https://support.apple.com/en-ca/HT204915), but can't hurt to check since this is all private-ish API.